File: backtestApp/views/result_manager.py
from PyQt5.QtWidgets import QWidget, QVBoxLayout, QTabWidget
import logging
from backtestApp.views.chart_view import ChartView
from backtestApp.views.stats_view import StatsView
from backtestApp.views.histogram_view_candle import HistogramViewCandle
from backtestApp.views.histogram_view import HistogramView
import time

class ResultManager:
    """Gestionnaire des vues de résultats du backtest."""
    
    def __init__(self, parent=None):
        """Initialise le gestionnaire avec une référence à l'application parent."""
        self.parent = parent
        self.views = {}
        
    def create(self):
        """Crée la zone d'affichage des résultats avec les différentes vues."""
        self.results_widget = QWidget()
        self.results_layout = QVBoxLayout(self.results_widget)
        
        # Créer les onglets pour séparer les différentes vues
        self.tab_widget = QTabWidget()
        
        # Initialiser les vues
        self.views['stats'] = StatsView(self.parent)
        self.views['chart'] = ChartView(self.parent)
        self.views['histogram'] = HistogramView(self.parent)
        # self.views['histogram_candle'] = HistogramViewCandle(self.parent)
        
        # Ajouter les vues aux onglets
        self.tab_widget.addTab(self.views['stats'].create(), "📊 Statistiques")
        self.tab_widget.addTab(self.views['chart'].create(), "📈 Graphiques")
        self.tab_widget.addTab(self.views['histogram'].create(), "📊 Histogramme PnL")
        # self.tab_widget.addTab(self.parent.histogram_view_candle.create(), "📊 Histogramme Bougies")
        
        # Ajouter le widget d'onglets au layout principal
        self.results_layout.addWidget(self.tab_widget)
        
        return self.results_widget
    
    def update_all(self, chart_data, stats):
        """Met à jour toutes les vues avec les résultats du backtest"""
        import logging
        
        # Déboguer les données reçues
        logging.debug(f"update_all appelé avec chart_data de taille {len(chart_data)} et stats: {type(stats)}")
        if '_trades' in stats:
            logging.debug(f"Nombre de trades: {len(stats['_trades'])}")
        
        try:
            # Mettre à jour chaque vue
            for view_name, view in self.views.items():
                logging.debug(f"Mise à jour de la vue: {view_name}")
                view.update(chart_data, stats)
            
            logging.info("Toutes les vues ont été mises à jour avec succès")
        except Exception as e:
            logging.exception(f"Erreur lors de la mise à jour des vues: {str(e)}")
    
    def set_current_tab(self, index):
        """Définit l'onglet actif."""
        if 0 <= index < self.tab_widget.count():
            self.tab_widget.setCurrentIndex(index)
            
    def get_tab_widget(self):
        """Retourne le widget d'onglets."""
        return self.tab_widget

File: backtest_app_cpp/src/views/result_manager.cpp
#include "result_manager.h"
#include "stats_view.h"
#include "chart_view.h"
#include "histogram_view.h"
#include <QDebug>
#include <QTime>

ResultManager::ResultManager(QObject* parent) :
    BaseView(parent), 
    m_resultsWidget(nullptr),
    m_resultsLayout(nullptr),
    m_tabWidget(nullptr),
    m_statsView(nullptr),
    m_chartView(nullptr),
    m_histogramView(nullptr)
{
    qDebug() << "ResultManager créé avec parent:" << parent;
}

ResultManager::~ResultManager()
{
    // Qt gère automatiquement la destruction des widgets enfants
}

QWidget* ResultManager::create(QWidget* parentWidget)
{
    QTime start = QTime::currentTime();
    
    // Créer le widget principal des résultats
    m_resultsWidget = new QWidget(parentWidget);
    m_resultsLayout = new QVBoxLayout(m_resultsWidget);
    
    // Créer les onglets pour séparer les différentes vues
    m_tabWidget = new QTabWidget();
    
    // Passer 'this' comme parent QObject pour les vues
    m_statsView = new StatsView(this);
    m_histogramView = new HistogramView(this);
    m_chartView = new ChartView(this);
    
    // Vérifier que les vues ont été créées correctement
    if (!m_statsView || !m_histogramView || !m_chartView) {
        qCritical() << "Erreur lors de la création des vues";
        return nullptr;
    }
    
    // Créer les widgets des vues en passant le widget parent approprié
    QWidget* statsWidget = m_statsView->create(m_resultsWidget);
    QWidget* histogramWidget = m_histogramView->create(m_resultsWidget);
    QWidget* chartWidget = m_chartView->create(m_resultsWidget);
    
    // Vérifier que les widgets ont été créés
    if (!statsWidget || !histogramWidget || !chartWidget) {
        qCritical() << "Erreur lors de la création des widgets de vue";
        return nullptr;
    }
    
    m_tabWidget->addTab(statsWidget, "📊 Statistiques");
    m_tabWidget->addTab(histogramWidget, "📊 Histogramme PnL");
    m_tabWidget->addTab(chartWidget, "📈 Graphiques");
    
    // Ajouter au map pour faciliter l'accès
    m_views["stats"] = m_statsView;
    m_views["histogram"] = m_histogramView;
    m_views["chart"] = m_chartView;
    
    // Ajouter le widget d'onglets au layout principal
    m_resultsLayout->addWidget(m_tabWidget);
    
    int elapsed = start.msecsTo(QTime::currentTime());
    qInfo() << "ResultManager::create() took" << elapsed << "ms";
    
    return m_resultsWidget;
}

void ResultManager::update(void* data, void* stats)
{
    updateAll(data, stats);
}

void ResultManager::clear()
{
    for (auto it = m_views.begin(); it != m_views.end(); ++it) {
        it.value()->clear();
    }
}

void ResultManager::updateAll(void* data, void* stats)
{
    QTime start = QTime::currentTime();
    
    qDebug() << "updateAll appelé avec des pointeurs opaques data et stats";
    
    try {
        // Définir l'ordre d'exécution pour éviter les conflits
        QStringList updateOrder = {"stats", "histogram", "chart"};
        
        for (const QString& viewName : updateOrder) {
            if (m_views.contains(viewName)) {
                qDebug() << "Mise à jour de la vue:" << viewName;
                m_views[viewName]->update(data, stats);
            }
        }
        
        qInfo() << "Toutes les vues ont été mises à jour avec succès";
    }
    catch (const std::exception& e) {
        qCritical() << "Erreur lors de la mise à jour des vues:" << e.what();
    }
    
    int elapsed = start.msecsTo(QTime::currentTime());
    qInfo() << "ResultManager::updateAll() took" << elapsed << "ms";
}

void ResultManager::setCurrentTab(int index)
{
    if (m_tabWidget && index >= 0 && index < m_tabWidget->count()) {
        m_tabWidget->setCurrentIndex(index);
    }
}

QTabWidget* ResultManager::getTabWidget() const
{
    return m_tabWidget;
}